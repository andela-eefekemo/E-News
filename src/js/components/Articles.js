import React from 'react';

const Article = (props) => {
  const { urlToImage, description, title, url } = props;
  return (
    <div className="row">
      <div className="col s12 m6">
        <div className="card">
          <div className="card-image">
            <img src={urlToImage} alt="Headline" />
            <span className="card-title">{title}</span>
          </div>
          <div className="card-content">
            <p>{description}</p>
          </div>
          <div className="card-action">
            <a href={url} target="_blank">View More</a>
          </div>
        </div>
      </div>
    </div>

  );
};

export default Article;

