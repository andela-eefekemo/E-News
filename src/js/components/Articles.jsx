import React from 'react';
import PropTypes from 'prop-types';

// Article Component
const Article = (props) => {
  const { urlToImage, description, title, url } = props;
  return (

    <div className="col s12 m6 l4">
      <div className="card card-panel hoverable">
        <div className="card-image">
          <img src={urlToImage} alt="Headline" />
          <span className="card-title">{title}</span>
        </div>
        <div className="card-content">
          <p className="truncate">{description}</p>
        </div>
        <div className="card-action">
          <a href={url} target="_blank" rel="noopener noreferrer">View More</a>
        </div>
      </div>
    </div>
  );
};

// Set Default Props
Article.defaultProps = {
  urlToImage: '',
  description: '',
  title: '',
  url: '',
};
// Set Props
Article.propTypes = {
  urlToImage: PropTypes.string,
  description: PropTypes.string,
  title: PropTypes.string,
  url: PropTypes.string,
};

export default Article;

