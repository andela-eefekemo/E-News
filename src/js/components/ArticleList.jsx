/* global localStorage*/
import React from 'react';

import PropTypes from 'prop-types';
import Nav from './layout/Nav.jsx';
import Footer from './layout/Footer.jsx';
import articlesStore from '../stores/articlesStore';
import * as NewsActions from '../actions/newsAction';
import Dropdown from './Dropdown.jsx';
import Articles from './Articles.jsx';

/**
 * @class ArticleList
 * @extends {React.Component}
 */
class ArticleList extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      articles: [],
      name: 'articles',
      currentSort: '',
    };
    this.article = this.getArticle.bind(this);
    this.sortsBy = [];
    this.name = '';
    this.id = '';
    this.change = this.onChange.bind(this);
  }
  /**
   * @method componentDidMount - Runs after the page has been rendered
   * @return {void}
   * Retrieves query values from the props and assigns them to instance variables
   * Makes an action call to get list of sources from the Api
   * Listens for a change event from the sourcesStore
   */
  componentDidMount() {
    const { location } = this.props;
    const { query } = location;
    this.sorts = query.sorts.split(',');
    this.sortsBy = this.sorts.map(sort => sort);
    this.name = query.name;
    this.id = query.id;
    NewsActions.getArticles(query.id, this.sorts[0]);
    articlesStore.on('changes', this.article);
  }
  /**
   * @method componentWillUnmount - Runs after component has been closed
   * @return {void}
   * Removes changes Listener from the articlesStore
   */
  componentWillUnmount() {
    articlesStore.removeListener('changes', this.article);
  }
  /**
   * @method onChange - Listens for an event and makes an action call to the Api
   * @param {event} event -Takes in an event parameter
   * @return {void}
   */
  onChange(event) {
    const value = event.target.value;
    NewsActions.getArticles(this.id, value);
    this.setState({ currentSort: value });
  }
  /**
   * @method getArticle - Sets the state of articles to data retrieve from articlesStore
   * @return {void}
   */
  getArticle() {
    this.setState({
      articles: articlesStore.getArticles(),
    });
  }
  render() {
    // maps through sortsBy array and passes props to dropdown component
    const sorted = this.sortsBy.map((sort, index) => (
      <Dropdown key={index} value={sort} text={sort} />
    ));
    // maps through articles array and passes props to Articles component
    const singleArticles = this.state.articles.map(article => (
      <Articles key={article.publishedAt} {...article} />
    ));
    return (
      <div>
        <Nav />
        <div className="container" id={this.state.name}>
          <div className="row">
            <div className="col m6">
              <h1>{this.name}</h1>
            </div>
            <div className="col m6">
              <div className="row">
                <div className="col m4">
                  <h6 htmlFor="sortsBy" className="sortBy">Sort By</h6>
                </div>
                <div className="col m8">
                  <select
                    className="browser-default input-field select"
                    onChange={this.change}
                    id="sortsBy"
                  >
                    {sorted}
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div className="row">
            {singleArticles}
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}

ArticleList.defaultProps = {
  location: {
    query: {
      sorts: 'top',
    },
  },
};

ArticleList.propTypes = {
  location: PropTypes.object,
};

export default ArticleList;
