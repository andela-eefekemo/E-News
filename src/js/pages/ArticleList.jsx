/* global localStorage*/
import React from 'react';

import PropTypes from 'prop-types';
import Nav from '../components/layout/Nav.jsx';
import Footer from '../components/layout/Footer.jsx';
import NewsStore from '../stores/newsStore';
import * as NewsActions from '../actions/newsAction';
import Dropdown from '../components/Dropdown.jsx';
import Articles from '../components/Articles.jsx';

/**
 * Headlines React Component
 */
class Headlines extends React.Component {
  constructor(props) {
    super(props);
    this.api = NewsStore.getArticles();
    this.state = {
      articles: this.api,
      name: 'articles',
    };
    this.article = this.getArticle.bind(this);
    this.sortsBy = [];
    this.name = '';
    this.id = '';
    this.change = this.onChange.bind(this);
  }
  // componentDidMount - Runs when component is loaded
  componentDidMount() {
    const { location } = this.props;
    const { query } = location;
    this.sorts = query.sorts.split(',');
    this.sortsBy = this.sorts.map(sort => sort);
    this.name = query.name;
    this.id = query.id;
    NewsStore.on('changes', this.article);
    NewsActions.getArticle(query.id, this.sorts[0]);
  }
  // componentWillUnMount - Runs when component is unloaded(changed)
  componentWillUnmount() {
    NewsStore.removeListener('changes', this.article);
  }
  // onChange - Runs when onChange event is called in select field
  onChange(event) {
    const value = event.target.value;
    NewsActions.getArticle(this.id, value);
  }
  // getArticle - Sets state of article
  getArticle() {
    this.setState({
      articles: NewsStore.getArticles(),
    });
  }
  render() {
    // maps through sortsBy array and passes props to dropdown component
    const sorted = this.sortsBy
      .map((sort, index) => <Dropdown key={index} value={sort} text={sort} />);
    // maps through articles array and passes props to Articles component
    const singleArticles = this.state.articles
      .map(article => <Articles key={article.publishedAt} {...article} />);
    return (
      <div>
        <Nav />
        <div className="container" id={this.state.name}>
          <div className="row">
            <div className="col m6">
              <h1>{this.name}</h1>
            </div>
            <div className="col m6">
              <div className="row">
                <div className="col m4">
                  <h6 htmlFor="sortsBy" className="sortBy">Sort By</h6>
                </div>
                <div className="col m8">
                  <select
                    className="browser-default input-field select"
                    onChange={this.change} id="sortsBy">
                    {sorted}
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div className="row">
            {singleArticles}
          </div>
        </div>
        <Footer />
      </div>
    );
  }
}
// sets default propstype
Headlines.defaultProps = {
  location: {
    query: {
      sorts: 'top',
    },
  },
};

Headlines.propTypes = {
  location: PropTypes.object,
};

export default Headlines;
