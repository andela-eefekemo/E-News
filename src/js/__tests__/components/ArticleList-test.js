import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer'

import ArticleList from '../../pages/ArticleList';
import Articles from '../../components/Articles.jsx';

test('Headline Component', () => {
  const component = toJson(mount(<ArticleList sorts="top"/>));
  expect(component).toMatchSnapshot();
});


test('Testing filter function', () =>{
  const onChange= jest.fn();
  const component = mount(<ArticleList onChange={onChange} />);
  expect(component.props().onChange).toBeDefined();
  const input = component.find('select');
  expect(input).toBeDefined();
  expect(component.state().currentSort).toBe('');
  input.simulate('change', {target: { value: 'top' }});
  expect(component.state().currentSort).toBe('top');
});