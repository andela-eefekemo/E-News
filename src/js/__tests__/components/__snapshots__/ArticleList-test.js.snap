// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Headline Component 1`] = `
<ArticleList
  location={
    Object {
      "query": Object {
        "sorts": "top",
      },
    }
  }
  sorts="top"
>
  <div>
    <Nav
      imageURL=""
      info={
        Object {
          "imageURL": "",
          "name": "",
        }
      }
      name=""
    >
      <div>
        <nav>
          <div
            className="nav-wrapper teal lighten-1"
          >
            <a
              className="brand-logo logo"
              href="#!"
            >
              E-News
            </a>
            <a
              className="button-collapse"
              data-activates="mobile-demo"
              href="#!"
            >
              <img
                alt="profile"
                className="pic"
                src=""
              />
            </a>
            <ul
              className="right hide-on-med-and-down"
            >
              <li>
                <IndexLink
                  to="/"
                >
                  <Link
                    onlyActiveOnIndex={true}
                    style={Object {}}
                    to="/"
                  >
                    <a
                      onClick={[Function]}
                      style={Object {}}
                    >
                      Home
                    </a>
                  </Link>
                </IndexLink>
              </li>
              <li>
                <a
                  className="dropdown-button"
                  data-activates="dropdown1"
                  href="#"
                >
                  <i
                    className="material-icons right"
                  >
                    arrow_drop_down
                  </i>
                </a>
              </li>
              <ul
                className="dropdown-content"
                id="dropdown1"
              >
                <li>
                  <img
                    alt="profile"
                    className="pic"
                    src=""
                  />
                </li>
                <li
                  className="divider"
                />
                <li>
                  <Link
                    onClick={[Function]}
                    onlyActiveOnIndex={false}
                    style={Object {}}
                    to="/logout"
                  >
                    <a
                      onClick={[Function]}
                      style={Object {}}
                    >
                      Logout
                    </a>
                  </Link>
                </li>
              </ul>
            </ul>
            <ul
              className="side-nav"
              id="mobile-demo"
            >
              <li>
                <Link
                  onlyActiveOnIndex={false}
                  style={Object {}}
                  to="/"
                >
                  <a
                    onClick={[Function]}
                    style={Object {}}
                  >
                    Home
                  </a>
                </Link>
              </li>
              <li
                className="divider"
              />
              <li>
                 
                <Link
                  onClick={[Function]}
                  onlyActiveOnIndex={false}
                  style={Object {}}
                  to="/logout"
                >
                  <a
                    onClick={[Function]}
                    style={Object {}}
                  >
                    Logout
                  </a>
                </Link>
              </li>
            </ul>
          </div>
        </nav>
      </div>
    </Nav>
    <div
      className="container"
      id="articles"
    >
      <div
        className="row"
      >
        <div
          className="col m6"
        >
          <h1 />
        </div>
        <div
          className="col m6"
        >
          <div
            className="row"
          >
            <div
              className="col m4"
            >
              <h6
                className="sortBy"
                htmlFor="sortsBy"
              >
                Sort By
              </h6>
            </div>
            <div
              className="col m8"
            >
              <select
                className="browser-default input-field select"
                id="sortsBy"
                onChange={[Function]}
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="row"
      />
    </div>
    <Footer>
      <div
        className="footer"
      >
        <p
          className="center-align"
        >
          Â© 2017 | E-News
        </p>
        <p
          className="center-align"
        >
          Developed by: Efekemo Eguonoghene
        </p>
      </div>
    </Footer>
  </div>
</ArticleList>
`;
